### PR TITLE
fix: Internal Sharding, this time fixed™

### LINKED ISSUE #2601: Bot silently disconnects from Discord.
**Please describe the problem you are having in as much detail as possible:**
Bot silently disconnects when left on for hours. The issue probably not related with my own code. Uploaded is the runtime when it occured.
[runtime.log](https://github.com/discordjs/discord.js/files/2107669/runtime.log)

**Further details:**

- discord.js version: latest 11.3.2
- node.js version: 8.11.2
- Operating system: Ubuntu 16.04.4 LTS (GNU/Linux 4.15.13-x86_64-linode106 x86_64)
- Priority this issue should have – please be realistic and elaborate if possible:

Minor. Restarting should make it work again but it's a pain to restart it every now and then.

- [ ] I found this issue while running code on a __user account__
- [ ] I have also tested the issue on latest master, commit hash:

### LINKED ISSUE #3021: Shard 0 gateway timing out after a few seconds
**Please describe the problem you are having in as much detail as possible:**

Ever since updating to a recent version of master branch, my shard 0 has been timing out with the gateway after a few seconds of functioning fine. I can reproduce this reliably each time I turn the bot on.

**Include a reproducible code sample here, if possible:**

My code hasn't changed other than updating dependencies. Here's the testing I did for each hash since a working version and the latest version:
```
BROKEN  8230255c68b94d68a4e8ffc559a98d08d1a08a7c
WORKING 28db5273708af039ef6a5387273d4407d55f178c
WORKING 840d22f8920f7648f97109d3df6f8a9b1da9f20f
WORKING 0bde9ca2c58785f2e1093698268aa9bdc525f5a5
WORKING 8a76cc5c725ffcfe8131c652de3d16be1f51439a
WORKING 8286d1a0fc3e86a10b7d5f977836b0e0fe611c27
....... 21ec03140f182b93ddc151e9edbdb6da8d8ae703
....... 5cbdf380289f0fd1d16330325dab71f8815396fc 
WORKING b5d5c699e65bf14a77065a75aabd24e4972988a9
```
There a few (>5) times where results were inconsistent, but that might be me screwing up my tests. If you really care about code, [this is my bot's code](https://github.com/dice-discord/bot/commit/02f59339cfa6c2364daa4c3b429c52ff1ea9be04) where I can confirm the problem was occurring.

Here are logs with debug mode on of my bot: https://hastebin.com/podokozise.log.

**Further details:**
- discord.js version: master
- Node.js version: `10.15.0`
- Operating system: Debian 9.5
- Priority this issue should have – please be realistic and elaborate if possible: low (I seem to be the only person experiencing it)

I've tried the following:
- deleting node_modules
- turning it off and on again
- switching computers
- resetting token

All of which had no effect. The only thing that seems to have any impact on the problem is changing which Discord.js version I use.

This issue has had me wracking my brain for the last few days, any help would be greatly appreciated.

- [x] I have also tested the issue on latest master, commit hash: 8230255c68b94d68a4e8ffc559a98d08d1a08a7c

### LINKED ISSUE #3053: Master Version Causing WebSocket Timeout
Hello, 

As of 2 weeks ago i was running a slightly outdated version of the master branch of discord.js, my bot with 8 shards was working perfectly fine and booting and rebooting with no problems. When i was performing updates a few days ago, i updated the discord.js master branch to latest and suddenly the shards kept throwing ``Error [WS_CONNECTION_TIMEOUT]: The connection to the gateway timed out.``

I tried extending the time between booting shards, making it 15 seconds, no wait for ready. I tried manually extending the waitforready timeout (required for me to use wait for ready) and still timed out. I tried extending the shards to 12 and it still timedout, only when i bumped my shard count all the way up to 16 did the shards finally boot. 

I tried turning off all my events as well and this did not stop it from hitting a timeout.

On a lark i downgraded from master to the stable branch and turned off my bot (simply flipped an if statement that prevents it from acknowledging the 'ready' state) and all 8 shards booted it pretty much the same fashion as they usually do no problems whatsoever.

I do not understand how this could be an issue nor can i really give any ideas where it might be caused, but i think i can firmly say that this is not a problem on my end. If you have a disagreement with that, please let me know I'm wiling to try any solution you give me.

Full Error:

{ Error [WS_CONNECTION_TIMEOUT]: The connection to the gateway timed out.
0|shardManager  |     at Timeout.setTimeout (/home/holo/gaius/node_modules/discord.js/src/client/Client.js:261:16)
0|shardManager  |     at ontimeout (timers.js:436:11)
0|shardManager  |     at tryOnTimeout (timers.js:300:5)
0|shardManager  |     at unrefdHandle (timers.js:520:7)
0|shardManager  |     at Timer.processTimers (timers.js:222:12) [Symbol(code)]: 'WS_CONNECTION_TIMEOUT' }

Note: This error happened on 3 of 8 shards and only on those shards i know had the most users (around 280k

**Further details:**
- discord.js version: Master
- Node.js version: 10.15.0
- Operating system: Ubuntu
- Priority this issue should have – please be realistic and elaborate if possible: I'm not good at delegating priorities, so just acknowledgement of existence is fine i suppose

- [ ] I have also tested the issue on latest master, commit hash:

### LINKED ISSUE #3083: WebSocket issues
**Please describe the problem you are having in as much detail as possible:**
Having multiples websocket/websocketshard issues after one of the lastest commit.

**Include a reproducible code sample here, if possible:**
```js
SyntaxError: Unexpected token t in JSON at position 1
    at JSON.parse (<anonymous>)
    at Object.exports.unpack.data [as unpack] (/home/koyamie/Koya/node_modules/discord.js/src/WebSocket.js:25:48)
    at WebSocketShard.onMessage (/home/koyamie/Koya/node_modules/discord.js/src/client/websocket/WebSocketShard.js:240:26)
    at WebSocket.onMessage (/home/koyamie/Koya/node_modules/discord.js/node_modules/ws/lib/event-target.js:120:16)
    at WebSocket.emit (events.js:189:13)
    at Receiver.receiverOnMessage (/home/koyamie/Koya/node_modules/discord.js/node_modules/ws/lib/websocket.js:742:20)
    at Receiver.emit (events.js:189:13)
    at Receiver.dataMessage (/home/koyamie/Koya/node_modules/discord.js/node_modules/ws/lib/receiver.js:408:14)
    at Receiver.getData (/home/koyamie/Koya/node_modules/discord.js/node_modules/ws/lib/receiver.js:347:17)
    at Receiver.startLoop (/home/koyamie/Koya/node_modules/discord.js/node_modules/ws/lib/receiver.js:133:22)
    at Receiver._write (/home/koyamie/Koya/node_modules/discord.js/node_modules/ws/lib/receiver.js:69:10)
    at doWrite (_stream_writable.js:415:12)
    at writeOrBuffer (_stream_writable.js:399:5)
    at Receiver.Writable.write (_stream_writable.js:299:11)
    at TLSSocket.socketOnData (/home/koyamie/Koya/node_modules/discord.js/node_modules/ws/lib/websocket.js:817:35)
    at TLSSocket.emit (events.js:189:13)
```

```js
TypeError: Cannot read property 'close' of null
    at WebSocketShard.sendHeartbeat (/home/koyamie/Koya/node_modules/discord.js/src/client/websocket/WebSocketShard.js:151:23)
    at Timeout.heartbeatInterval.manager.client.setInterval [as _onTimeout] (/home/koyamie/Koya/node_modules/discord.js/src/client/websocket/WebSocketShard.js:175:73)
    at listOnTimeout (timers.js:324:15)
    at processTimers (timers.js:268:5)
```

**Further details:**
- discord.js version: lastest master
- Node.js version: 11.4.0
- Operating system: Debian 9
- Priority this issue should have – please be realistic and elaborate if possible:

- [ ] I have also tested the issue on latest master, commit hash:

### LINKED ISSUE #3085: Internal shard stuck on 'Connecting' status
**Please describe the problem you are having in as much detail as possible:**

My websocket shards disconnected (code: 1000), and then reconnected - but one never reconnected and was just stuck on a status of "1" (which is the code for "CONNECTING" ). It emitted the debug log saying that its connecting to the gateway but nothing happened to the shard afterward. 

I was getting "Tried to send packet" errors from the shard when trying to update presence/status, as no websocket is available for it. 

It didn't reconnect or change by itself (i waited 10 minutes), calling `.connect()` on it instantly fixed it and reconnected it.

I'm assuming this is a bug, the solution seems to be to try to reconnect if its taking too long with a timeout?

**Include a reproducible code sample here, if possible:**
N/A

**Further details:**
- discord.js version: master https://github.com/discordjs/discord.js/commit/8910fed7298d5fb3d9b97d979bd1f0e0bd901145
- Node.js version: 10.14.2
- Operating system: ubuntu 16.04.4
- Priority: Low

### PR DESCRIPTION
**Pull request by vladfrangu**
**Please describe the changes this PR makes and why it should be merged:**

This PR fixes all the internal sharding issues (fingers crossed). To put it simply, this closes #3053, closes #3085, closes #3021, closes #3083 (or should, this one I'm not 100% on) and closes #2601 (should, but we can't just cause an outage to test out)

## But Vlad.. Didn't you say the same in #2976?

I did! You are correct. Buuut, development testing is nothing compared to real life testing, and as shown, people have experienced issues with IS, even after the mentioned PR got merged (although far less often!).

## So, how do we know this PR solves this?

Short answer: we don't.

Long answer, I want your help. If you had issues with IS, or are just willing to give a helping hand, please install this PR (`npm i vladfrangu/discOwOd.js#internal-sharding-2.0` or `yarn add vladfrangu/discOwOd.js#internal-sharding-2.0`) and run it. Logging the `debug` event on the client will be required, how you do it is up to you, but those logs can help in aid fixing issues you might encounter.

I'll also be working on the [Discord.JS sharding test suite](https://github.com/vladfrangu/discordjs-sharding-testing-suite) (at the time of writing, it's midnight and I can't work on this right now, but soon).
The plan is to implement as many different sharding ways people use and run this PR (or any for that matter) to make sure the end results are as expected. This is still, a dream, but I will work on this and let you know as soon as possible 🤞 

### What can I do if this PR solved my issues?

Approve it! That's the best way to show it worked perfectly for you (but do it after a more long period of testing)

## And for the technical peeps out there, what changed?

- Client.js
	- Made the login function handle just basic checks and presence parsing, leaving the main work on the WebSocketManager
		- Keeps everything in one place, easier to maintain
	- Made options.shards ALWAYS end up as an array to keep code simpler (was Gus's request)
	- Add 4 new client events
		- `shardDisconnected` (replaces `disconnect`), emitted when a shard will never reconnect again 
		- `shardError` emitted when a shard encounters any error
			- To discuss if this should remain or if the ERROR event should be used instead
		- `shardReconnecting` emitted when a shard is getting queued up for a reconnect
		- `shardResumed` emitted when a shard successfully resumes (not ready's)
- Constants.js
	- Renamed READY -> CLIENT_READY (since it's the only place it's used)
	- Add the 4 new client events
	- Added 3 new ShardEvents
		- Used internally by the WebSocketShard to emit consistent events that the Manager will process (yes this could've been hardcoded, but I like it cleaner)
- WebSocketManager.js
	- Added a totalShards property which is used in the checkReady function
	- Made `shardQueue` and `packetQueue` hidden by default (if you log the class)
	- `expectingClose` -> `destroyed`
	- `isReconnectingShards` -> `reconnecting`
	- Made the debug function take shard as an optional parameter, to be used in the debug messages
	- Made the connect function handle everything, from fetching gateway information to checking if the token is valid, creating the shard instances and initiating the connect process
	- Renamed `create` to `createShards` and implemented more logic to it
		- Attaches event handles on the shard if they weren't already
		- awaits! the connection of a shard and handles the error appropriately
		- Handles the session limits accordingly
	- Made the `reconnect` function re-use the `createShards` one, while still handling session limits AND token invalidation checks
- WebSocketShard.js
	- Made the `ratelimit`, `connection`, `inflate`, `helloTimeout` (more on this later) and `eventsAttached` not visible by default
	- Made the `connect` function return a Promise that resolves when the shard gets ready / resumes and rejects if it closes
	- `onOpen` sets the shard's status to `NEARLY`, to indicate we're almost READY
	- Simplified `onClose` by just emitting the close event and letting the manager handle the rest
	- Implement a HELLO payload timeout
		- This should prevent any dead or hanging shards that never connect, as well as ensure we're.. actually connected
	- Add an `important` parameter to `send`
		- Used only by important functions which need to send the payload NOW, not wait for the entire queue to drain (say, identify before presence updates)
	- Minor touch ups to the `destroy` function
- index.d.ts
	- I really can't be bothered to explain everything I did, so I'll just say I added everything about the classes and events. Happy? 

Have fun reviewing and testing! 🎉 🍰 

**Status**
- [x] Code changes have been tested against the Discord API, or there are no code changes
- [x] I know how to update typings and have done so, or typings don't need updating

**Semantic versioning classification:**  
- [x] This PR changes the library's interface (methods or parameters added)
  - [x] This PR includes breaking changes (methods removed or renamed, parameters moved or removed)
- [ ] This PR **only** includes non-code changes, like changes to documentation, README, etc.

### CODE REVIEW

File: src/client/websocket/WebSocketManager.js [hunk 1]
```diff
@@ -96,121 +109,159 @@ class WebSocketManager {
...
-    if (remaining !== 0) return true;
-    return reset_after;
-  }
+  async connect() {
+    const invalidToken = new Error(WSCodes[4004]);
+    const {
+      url: gatewayURL,
+      shards: recommendedShards,
+      session_start_limit: sessionStartLimit,
+    } = await this.client.api.gateway.bot.get().catch(() => { throw invalidToken; });
```
[the remark this answers is not in the document]
Comment (vladfrangu) [reply src/client/websocket/WebSocketManager.js]: While I agree 100% with you, in my opinion the REST side of the library should handle 5xx gracefully, and shouldn't throw... There's a retries property again too.. Discussion later

**Comment by vadremix** [comment]
Is it reasonable to use internal sharding in conjunction with the shard manager? E.g. 3 shards to a process

File: src/errors/Messages.js [hunk 2]
```diff
@@ -4,12 +4,12 @@ const { register } = require('./DJSError');
 
 const Messages = {
   CLIENT_INVALID_OPTION: (prop, must) => `The ${prop} option must be ${must}`,
+  CLIENT_INVALID_PROVIDED_SHARDS: 'None of the shards you provided were valid.',
```
Comment (SpaceEEC) [inline src/errors/Messages.js]: `None of the provided shards were valid.` would be more consistent ~~most of~~ the other error messages.

**Comment by rumblefrog** [comment]
Can confirm I no longer experience the same as #2976 

Thanks

**Comment by xonrsoftware** [comment]
Still appears silent disconnect. Just now two bot's went silently offline.
`discord.js@12.0.0-dev (github:vladfrangu/discOwOd.js#cd8747c06899ce938addac13f893921c3fb7f902)`
Will update to last commit and check again.

**Comment by vladfrangu** [comment]
Hey @xonrsoftware, thank you for reporting that you've encountered an issue with this PR!

If, per chance, the latest commits do not solve the silent disconnects, please provide a log file of what the `debug` event emitted. (`client.on('debug', message => {})`) 

That will aid me into figuring out where and what happened before they silently disconnected, in order to fix it ❤️

**Pull request merged**